cosmwasm/contracts/crosschain-registry/src/execute.rs

```diff
@@ -299,6 +317,104 @@ pub fn chain_to_prefix_operations(
     Ok(response)
 }
 
+// Struct for input data for a single chain to authorized address operation
+#[cw_serde]
+pub struct AuthorizedAddressInput {
+    pub operation: Operation,
+    pub source_chain: String,
+    pub addr: Addr,
+}
+
+pub fn authorized_address_operations(
+    deps: DepsMut,
+    sender: Addr,
+    operation: Vec<AuthorizedAddressInput>,
+) -> Result<Response, ContractError> {
+    let response = Response::new();
+    for operation in operation {
+        let addr = operation.addr;
+        let source_chain = operation.source_chain.to_lowercase();
+
+        // Only contract governor can call by-chain permission CRUD operations
+        check_is_contract_governor(deps.as_ref(), sender.clone())?;
+
+        match operation.operation {
+            Operation::Set => {
+                let authorized_addr = AUTHORIZED_ADDRESSES
+                    .may_load(deps.storage, &source_chain)
+                    .unwrap_or_default();
+
+                if authorized_addr.eq(&Some(addr.clone())) {
+                    return Err(ContractError::CustomError {
+                        msg: format!(
+                            "{} is already an authorized address for source chain {}",
+                            addr, source_chain
+                        ),
+                    });
+                }
+
+                AUTHORIZED_ADDRESSES.save(deps.storage, &source_chain, &addr)?;
+                response.clone().add_attribute(
+                    "set_authorized_address",
+                    format!("{}-{}", source_chain, addr),
+                );
+            }
+            Operation::Change => {
+                let authorized_addr = AUTHORIZED_ADDRESSES
+                    .load(deps.storage, &source_chain)
+                    .map_err(|_| ContractError::CustomError {
+                        msg: format!(
+                            "No authorized address found for source chain {}",
+                            source_chain
+                        ),
+                    })?;
+
+                if !authorized_addr.eq(&addr) {
```

Comment:
I don't think this is right. We expect authorized_addr != addr, since we're changing the aithorized addr and setting it to a new addr

Reply:
You are right! Fixed it here, but will likely change a little more and add respective tests as per our offline discussion

https://github.com/osmosis-labs/osmosis/pull/4553/commits/ad6814a2983c439fe5ca7aeb69e0a9746ed493ac